### PR TITLE
Fix publishing documentation

### PR DESCRIPTION
**Pull request by Avasam**
Since the help files aren't shipped with the wheel in 307, the `docs` workflow fails.

Given the catch-22 nature of generating docs at the moment, and not wanting to create extra builds when not needed. I kept the PyPI download. But regerenate a .chm from source. (the help file generation honestly isn't that long anyway).

I also added doc generation as a CI test to sanity test that a PR won't break the docs publishing after a tag is created.

This PR makes the following changes:
- Generate a new .chm file in docs workflow since it's no longer shipped in wheels
- Pass the current python executable path to nmake instead of hardcoding a specific version
- Remove the redundant zip step, since https://github.com/actions/upload-artifact#upload-an-entire-directory already does it
- Update GitHub actions in `docs` workflow (also done in #2376)
- PR validation that doc generation still works when changes are made in relevant files

There are no technical reasons preventing us from still shipping the help files (`PyWin32.chm`), other than maybe their size (see https://github.com/mhammond/pywin32/issues/1380#issuecomment-2019333366). If it's a long-term choice, then let's also remove the dead setup code to register them. (we could ship a web link file instead) Otherwise I can add it back to the wheel as part of the automated process. Either way that'll be a different PR.

This incidentally works towards #394

### CODE REVIEW

File: AutoDuck/common_top.mak [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
-HC		  = hcw /a /c /e
-HHC		  = hhc
-PYTHON   = py -2.7
+ADLOG     = /R help "/L$(GENDIR)\$(TARGET).HLOG" /N
+ADHTML    = /R html "/G$(GENDIR)\$(TARGET).DUMP" "/C$(GENDIR)\$(TARGET).HHLOG" "/I$(GENDIR)\$(TARGET).IDX" "/O$(GENDIR)\$(TARGET).HTML" /D "title=$(TITLE)"
+ADHLP     = /R help "/C$(GENDIR)\$(TARGET).HLOG" "/O$(GENDIR)\$(TARGET).RTF" /D "title=$(TITLE)"
+ADDOC     = /RD "/O$(GENDIR)\$(TARGET).DOC" /D "doc_header=$(DOCHDR)"
+ADTAB     = 8
+HC        = hcw /a /c /e
+HHC       = hhc
```
Comment (Avasam) [inline AutoDuck/common_top.mak:12]: If you still want this to be runnable w/o the python script, I think this should work to add a default executable location:
```suggestion
HHC       = hhc
PYTHON    = python
```

**Comment by Avasam** [comment]
Ah, another catch-22. I feel like this should be feasible to resolve.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -43,6 +43,13 @@ jobs:
         run: |
           python setup.py install --user
 
+      # This needs to happen *after* installing pywin32 since
+      # AutoDuck/py2d.py currently relies on runtime imports for introspection
+      # This isn't included in the wheel (TODO: could we?)
+      # and only servces as a PR test for the docs.yaml workflow
+      - name: Generate PyWin32.chm help file
+        run: python AutoDuck/make.py
+
```
Comment (Avasam) [inline .github/workflows/main.yml:52]: If we merged both workflows, we could push these as artifacts, then no need to download from PyPI

File: .github/workflows/docs.yml [hunk 3]
```diff
@@ -3,15 +3,18 @@ name: docs
...
-    # https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet
+      # https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet
       - b[0-9][0-9][0-9]
       - publish-docs*
   push:
     branches:
       - main
+  pull_request: # Temporary just for test
+    branches:
+      - main
```
Comment (Avasam) [inline .github/workflows/docs.yml:14]: I added this for the sake of testing in my PR, but if we keep this we wouldn't need to add a test in `.github/workflows/main.yml`

**Review by mhammond — APPROVED** [review]
Sorry for the delay, this is great, thanks!

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -43,6 +43,13 @@ jobs:
         run: |
           python setup.py install --user
 
+      # This needs to happen *after* installing pywin32 since
+      # AutoDuck/py2d.py currently relies on runtime imports for introspection
+      # This isn't included in the wheel (TODO: could we?)
```
Comment (Avasam) [inline .github/workflows/main.yml:48]: > TODO: could we?

Open question on setuptools discussion: https://github.com/pypa/setuptools/discussions/4685

Comment (Avasam) [inline .github/workflows/main.yml:48]: > TODO: could we?

Open question on setuptools discussion: https://github.com/pypa/setuptools/discussions/4685

**Pull request merged**